File: src/algorithm_java/leetcode/Problem234.java
package algorithm.leetcode;

public class Problem234 {

    public static boolean isPalindrome(ListNode head) {
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        if (fast != null)
            slow = slow.next;

        ListNode p = slow;
        ListNode pre = null;
        ListNode next = null;
        ListNode reverseHead = null;
        while (p != null) {
            next = p.next;
            if (next == null)
                reverseHead = p;
            p.next = pre;
            pre = p;
            p = next;
        }

        p = reverseHead;
        ListNode q = head;
        while (p != null) {
            if (p.val != q.val)
                return false;
            p = p.next;
            q = q.next;
        }
        return true;
    }

    public static void main(String[] args) {
        ListNode a = new ListNode(1);
        ListNode b = new ListNode(2);
        ListNode c = new ListNode(2);
        ListNode d = new ListNode(1);

        a.next = b;
        b.next = c;
        c.next = d;
        isPalindrome(a);
    }

}

File: src/algorithm_cpp/problem234.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
	if (head == nullptr || head->next == nullptr) return true;
	ListNode *fast = head, *slow = head;
	while(fast->next != nullptr) {
	    slow = slow->next;
	    fast = fast->next;
	    if (fast->next != nullptr) fast = fast->next;
	}
	ListNode *next = slow->next, *nextnext = nullptr;
    // 加上这行就能过 很奇怪 leetcode bug？
    slow->next = nullptr;
	while(next != nullptr) {
	    nextnext = next->next;
	    next->next = slow;
	    slow = next;
	    next = nextnext;
	}
    fast = head;
    bool flag = true;
	while (true) {
	    if (slow->val != fast->val) {flag = false; break;}
        if (fast->next == slow) break;
        if (fast == slow) break;
        fast = fast->next;
        slow = slow->next;
        
	}
	return flag;
    }
};

